prometheus/histogram.go

```diff
@@ -1575,3 +1593,152 @@ func addAndResetCounts(hot, cold *histogramCounts) {
 	atomic.AddUint64(&hot.nativeHistogramZeroBucket, atomic.LoadUint64(&cold.nativeHistogramZeroBucket))
 	atomic.StoreUint64(&cold.nativeHistogramZeroBucket, 0)
 }
+
+type nativeExemplars struct {
+	exemplars                       map[*dto.Exemplar]*totalItem
+	nativeHistogramExemplarTTL      time.Duration
+	nativeHistogramMaxExemplarCount uint32
+	timeList                        *timeItem
+	logarithmList                   *logarithmItem
```

Comment:
While the algorithm seems to do the right thing, and your implementation might be the most efficient one in terms of big-O runtime, I do believe that it has a lot of baseline overhead, with two linked list and a map and many many pointers all over the place. Pointers have to be traversed, and they also take a lot of memory, which matters because an instrumented program might create many histograms.

Given that only a moderate number of exemplars make sense fundamentally (because that's the whole idea of exemplars to just present a few example observations), I'm pretty sure that an approach with a single slice will be more efficient both in terms of actual runtime and memory footprint.

So maybe have a single slice of size `nativeHistogramMaxExemplarCount` and sort it by the observed value. You can iterate through all ~10 elements once, searching for the oldest element and for the two closest ones at the same time. That should be super fast in practice. Insertion into the slice requires moving all the elements above by one, which is something computers can do super fast. (See https://go.dev/wiki/SliceTricks for trick to do it fast.)

Reply:
> While the algorithm seems to do the right thing, and your implementation might be the most efficient one in terms of big-O runtime, I do believe that it has a lot of baseline overhead, with two linked list and a map and many many pointers all over the place. Pointers have to be traversed, and they also take a lot of memory, which matters because an instrumented program might create many histograms.
> 
> Given that only a moderate number of exemplars make sense fundamentally (because that's the whole idea of exemplars to just present a few example observations), I'm pretty sure that an approach with a single slice will be more efficient both in terms of actual runtime and memory footprint.
> 
> So maybe have a single slice of size `nativeHistogramMaxExemplarCount` and sort it by the observed value. You can iterate through all ~10 elements once, searching for the oldest element and for the two closest ones at the same time. That should be super fast in practice. Insertion into the slice requires moving all the elements above by one, which is something computers can do super fast. (See https://go.dev/wiki/SliceTricks for trick to do it fast.)

I understand your point, I think this is better since the number of exemplars may be only around 10, I will change the code like you said. @beorn7

Reply:
@beorn7 
I already changed the implementation from linked list into slice, but I did not keep the slice sorted as you suggested, here are my thoughts about this  

1. We should always iterate all exemplars to get the exemplar with oldest timestamp, so the time efficient should always be O(n)
2. If we do not keep the slice sorted, we can just change the exemplar in-place with new exemplar, so there is not need to adjust the slice with like  https://go.dev/wiki/SliceTricks does, which makes the code easier.

WDYT?

BTY, I will fix your other reviews when we can agree with this implementation.

Reply:
See my comment below. I think we don't want to find the exemplar closest to the new exemplar, but the pair of exemplars closest to each other (including the new exemplar) and then delete the older of the two in the couple.